Feature/crud cost center

markolu opened the pull request.

lumdcs committed:
added entity costCenter

lumdcs committed:
created DTOs and controller

lumdcs committed:
Implemented controller and service layer

lumdcs committed:
Added exceptions

lumdcs committed:
added exception CostCenterNotPresent

lumdcs committed:
moved exception checking to service layer

lumdcs committed:
Added mapper and moved mapping to controller layer

lumdcs committed:
Removed CostCenterMapperInterface

lumdcs committed:
Code formatter

lumdcs committed:
Fixed controller and service attributes and moved CostCenterService and CostCenterServiceImplementation to service package

megicivovic commented on travel/src/main/java/com/tx/travel/controller/CostCenterController.java:
```diff
@@ -0,0 +1,112 @@
+import org.springframework.web.server.ResponseStatusException;
+
+@RestController
+@RequestMapping("/api/cost-centers")
```
have a look at this [comment](https://github.com/tx-services-Internship/travel-java-backend-monolith-spyro/pull/4#discussion_r1031729324)

megicivovic commented on travel/src/main/java/com/tx/travel/controller/CostCenterController.java:
```diff
@@ -0,0 +1,112 @@
+  private final CostCenterMapperImplementation costCenterMapper;
+
+  public CostCenterController(
+      final CostCenterServiceImplementation costCenterService,
```
here should be interface `CostCenterService`

megicivovic commented on travel/src/main/java/com/tx/travel/controller/CostCenterController.java:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  @DeleteMapping("/{id}")
+  public ResponseEntity<?> deleteCostCenterById(@PathVariable("id") Long id) {
```
maybe you can remove the wildcard since you no it will be with no content

megicivovic commented on travel/src/main/java/com/tx/travel/service/exception/CostCenterCodeAlreadyExists.java:
```diff
@@ -0,0 +1,13 @@
+@ResponseStatus(code = HttpStatus.CONFLICT, reason = "Already exists")
+public class CostCenterCodeAlreadyExists extends RuntimeException {
+
+  public CostCenterCodeAlreadyExists(final String code) {
```
```suggestion
  public CostCenterCodeAlreadyExistsException(final String code) {
```

megicivovic commented on travel/src/main/java/com/tx/travel/service/exception/CostCenterNotPresent.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_FOUND, reason = "Value not present")
+public class CostCenterNotPresent extends RuntimeException {
```
```suggestion
public class CostCenterNotPresentException extends RuntimeException {
```

megicivovic commented on travel/src/main/java/com/tx/travel/mapper/CostCenterMapperImplementation.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class CostCenterMapperImplementation {
```
```suggestion
public class CostCenterMapper {
```
since there is no interface, it doen't need to have "implementation" in the name

megicivovic reviewed (changes requested):
Couple of cosmetic changes only 🧹  God work on exception handling and mapping!

lumdcs committed:
Fixed import in TravelController

lumdcs committed:
Fixed exception and mapper names, controller constructor, added apiPrefix to properties, changed hibernate ddl-auto to validate

megicivovic commented on travel/src/main/java/com/tx/travel/controller/CostCenterController.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  @PostMapping()
+  public ResponseEntity<CostCenterResponse> saveCostCenter(
```
maybe name it createCostCenter

megicivovic reviewed (commented):
Amazing work! 🚀  Just add that test please!

lumdcs committed:
Added service layer tests in class CostCenterServiceImplementationTest,
added maven surefire plugin,
added 3 mockito-core, mockito-junit-jupiter and junit-jupiter-api dependencies,
renamed method saveCostCenter in CostCenterController to createCostCenter

megicivovic commented on travel/src/test/java/com/tx/travel/service/CostCenterServiceImplementationTest.java:
```diff
@@ -0,0 +1,372 @@
+
+    String actualMessage = exception.getMessage();
+
+    assertEquals(expectedMessage, actualMessage);
```
nice! 💯

megicivovic reviewed (commented):
This looks really good! lets discuss in daily if there is more to be done 😉

petkocfc commented on pom.xml:
```diff
@@ -58,6 +58,24 @@
+            <dependency>
+                <groupId>org.mockito</groupId>
+                <artifactId>mockito-core</artifactId>
+                <version>4.5.1</version>
```
would be nice to see this as property like mockito-core.version or similar but for now it's ok to stay like this
